presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregationIntoTableScan.java

```diff
@@ -0,0 +1,214 @@
+                .collect(toImmutableList());
+
+        Optional<AggregationPushdownResult<TableHandle>> aggregationPushdownResult = metadata.applyAggregation(
+                context.getSession(),
+                tableScan.getTable(),
+                aggregateFunctions,
+                assignments,
+                groupByColumns.isEmpty() ? ImmutableList.of() : ImmutableList.of(groupByColumns));
+
+        if (!aggregationPushdownResult.isPresent()) {
+            return Result.empty();
+        }
+
+        AggregationPushdownResult<TableHandle> result = aggregationPushdownResult.get();
+
+        TableHandle handle = result.getHandle();
+
+        // The new scan outputs and assignments are union of existing and the ones returned by connector pushdown.
+        ImmutableList.Builder<Symbol> newScanOutputs = new ImmutableList.Builder<>();
+        newScanOutputs.addAll(tableScan.getOutputSymbols());
+
+        ImmutableMap.Builder<Symbol, ColumnHandle> newScanAssignments = new ImmutableMap.Builder<>();
+        newScanAssignments.putAll(tableScan.getAssignments());
+
+        Map<String, Symbol> variableMappings = new HashMap<>();
+
+        for (Assignment assignment : result.getAssignments()) {
+            Symbol symbol = context.getSymbolAllocator().newSymbol(assignment.getVariable(), assignment.getType());
+
+            newScanOutputs.add(symbol);
+            newScanAssignments.put(symbol, assignment.getColumn());
+            variableMappings.put(assignment.getVariable(), symbol);
+        }
+
+        List<Expression> newPartialProjections = result.getProjections().stream()
+                .map(expression -> ConnectorExpressionTranslator.translate(expression, variableMappings, new LiteralEncoder(metadata)))
+                .collect(toImmutableList());
+
+        Assignments.Builder assignmentBuilder = Assignments.builder();
+        IntStream.range(0, aggregationOutputSymbols.size())
+                .forEach(index -> assignmentBuilder.put(aggregationOutputSymbols.get(index), newPartialProjections.get(index)));
+
+        // projections assignmentBuilder should have both agg and group by so we add all the group bys as symbol references
+        groupingSets.getGroupingKeys().stream()
+                .forEach(groupBySymbol -> assignmentBuilder.put(groupBySymbol, groupBySymbol.toSymbolReference()));
+
+        return Result.ofPlanNode(
+                new ProjectNode(
+                        context.getIdAllocator().getNextId(),
+                        TableScanNode.newInstance(
+                                context.getIdAllocator().getNextId(),
+                                handle,
+                                newScanOutputs.build(),
+                                newScanAssignments.build()),
+                        assignmentBuilder.build()));
+    }
+
+    private List<AggregateFunction> toAggregateFunction(List<AggregationNode.Aggregation> aggregations)
```

Comment:
Once you apply all the suggestions above and below, this method can simplify to:

```java
private AggregateFunction toAggregateFunction(AggregationNode.Aggregation aggregation)
{
    Signature signature = aggregation.getResolvedFunction().getSignature();

    ImmutableList.Builder<ConnectorExpression> arguments = new ImmutableList.Builder<>();
    for (int i = 0; i < aggregation.getArguments().size(); i++) {
        SymbolReference argument = (SymbolReference) aggregation.getArguments().get(i);
        arguments.add(new Variable(argument.getName(), metadata.getType(signature.getArgumentTypes().get(i))));
    }

    Optional<OrderingScheme> orderingScheme = aggregation.getOrderingScheme();
    Optional<List<Map.Entry<String, SortOrder>>> sortBy = orderingScheme.map(orderings ->
            orderings.getOrderBy().stream()
                    .map(orderBy -> new AbstractMap.SimpleEntry<>(
                            orderBy.getName(),
                            SortOrder.valueOf(orderings.getOrderings().get(orderBy).name())))
                    .collect(toImmutableList()));

    return new AggregateFunction(
            signature.getName(),
            metadata.getType(signature.getReturnType()),
            arguments.build(),
            sortBy.orElse(ImmutableList.of()),
            aggregation.isDistinct(),
            aggregation.getFilter().map(Symbol::getName));
}
```

Reply:
what about `aggregation.getMask()`? we should verify aggregation.getMask() is not present (yet).

Reply:
Done, added check for masks in pattern.